File: step_11_heaps/11.2_medium/IB121_distinct_numbers_in_window/Solution.java
/**
 * Problem: Distinct Numbers in Window
 * Difficulty: MEDIUM | XP: 25
 * Source: InterviewBit
 *
 * Given array A of N integers and window size K, return an array of size N-K+1
 * where each element is the count of distinct integers in the corresponding window.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE - Recount distinct for each window
    // Time: O(N * K)  |  Space: O(K)
    // ============================================================
    public static int[] bruteForce(int[] A, int k) {
        if (k > A.length) return new int[0];
        int n = A.length;
        int[] result = new int[n - k + 1];
        for (int i = 0; i <= n - k; i++) {
            Set<Integer> set = new HashSet<>();
            for (int j = i; j < i + k; j++) set.add(A[j]);
            result[i] = set.size();
        }
        return result;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - Sliding window with HashMap frequency count
    // Time: O(N)  |  Space: O(K)
    // Maintain frequency map. When count drops to 0, remove from map (= fewer distinct).
    // ============================================================
    public static int[] optimal(int[] A, int k) {
        if (k > A.length) return new int[0];
        int n = A.length;
        int[] result = new int[n - k + 1];
        Map<Integer, Integer> freq = new HashMap<>();

        // Initialize first window
        for (int i = 0; i < k; i++)
            freq.merge(A[i], 1, Integer::sum);

        result[0] = freq.size();

        for (int i = k; i < n; i++) {
            // Add new element
            freq.merge(A[i], 1, Integer::sum);
            // Remove outgoing element
            int out = A[i - k];
            freq.put(out, freq.get(out) - 1);
            if (freq.get(out) == 0) freq.remove(out);
            result[i - k + 1] = freq.size();
        }
        return result;
    }

    // ============================================================
    // APPROACH 3: BEST - Sliding window, same O(N) but with array freq
    // Time: O(N)  |  Space: O(max_val) or O(K) with HashMap
    // Same as optimal; use array if values are bounded, else HashMap.
    // For generality, identical to optimal but highlights key insight:
    // distinct count only changes at boundaries of the window.
    // ============================================================
    public static int[] best(int[] A, int k) {
        return optimal(A, k); // identical time/space; refer to optimal's logic
    }

    public static void main(String[] args) {
        System.out.println("=== Distinct Numbers in Window ===");

        int[] A1 = {1, 2, 1, 3, 4, 3};
        int k1 = 3;
        System.out.println("A1 k=3 brute: " + Arrays.toString(bruteForce(A1, k1)));   // [2,3,3,3]
        System.out.println("A1 k=3 opt:   " + Arrays.toString(optimal(A1, k1)));
        System.out.println("A1 k=3 best:  " + Arrays.toString(best(A1, k1)));

        int[] A2 = {1, 1, 1, 1};
        System.out.println("A2 k=2 brute: " + Arrays.toString(bruteForce(A2, 2)));     // [1,1,1]
        System.out.println("A2 k=2 opt:   " + Arrays.toString(optimal(A2, 2)));
    }
}

File: step_11_heaps/11.2_medium/IB121_distinct_numbers_in_window/solution.py
"""
Problem: Distinct Numbers in Window
Difficulty: MEDIUM | XP: 25
Source: InterviewBit

Given array A of N integers and window size K, return array of size N-K+1
where result[i] = number of distinct integers in A[i..i+K-1].
"""
from typing import List
from collections import defaultdict


# ============================================================
# APPROACH 1: BRUTE FORCE - Use set for each window
# Time: O(N * K)  |  Space: O(K)
# ============================================================
def brute_force(A: List[int], k: int) -> List[int]:
    """Count distinct in each window by building a new set each time."""
    n = len(A)
    if k > n:
        return []
    return [len(set(A[i:i+k])) for i in range(n - k + 1)]


# ============================================================
# APPROACH 2: OPTIMAL - Sliding window with frequency HashMap
# Time: O(N)  |  Space: O(K)
# Add new element, remove outgoing. Track distinct count via map size.
# ============================================================
def optimal(A: List[int], k: int) -> List[int]:
    """
    Maintain a frequency dict for the current window.
    When freq[x] drops to 0, del it (so len(freq) == distinct count).
    """
    n = len(A)
    if k > n:
        return []
    freq = defaultdict(int)
    result = []

    for i in range(k):
        freq[A[i]] += 1
    result.append(len(freq))

    for i in range(k, n):
        freq[A[i]] += 1
        out = A[i - k]
        freq[out] -= 1
        if freq[out] == 0:
            del freq[out]
        result.append(len(freq))

    return result


# ============================================================
# APPROACH 3: BEST - Counter-based sliding window (Pythonic)
# Time: O(N)  |  Space: O(K)
# Use collections.Counter for clean code; same asymptotic complexity.
# ============================================================
def best(A: List[int], k: int) -> List[int]:
    """Counter-based sliding window for distinct count."""
    from collections import Counter
    n = len(A)
    if k > n:
        return []
    window = Counter(A[:k])
    result = [len(window)]
    for i in range(k, n):
        window[A[i]] += 1
        window[A[i - k]] -= 1
        if window[A[i - k]] == 0:
            del window[A[i - k]]
        result.append(len(window))
    return result


if __name__ == "__main__":
    print("=== Distinct Numbers in Window ===")
    tests = [
        ([1, 2, 1, 3, 4, 3], 3, [2, 3, 3, 3]),
        ([1, 1, 1, 1], 2, [1, 1, 1]),
        ([1, 2, 3, 4, 5], 3, [3, 3, 3]),
        ([1], 1, [1]),
    ]
    for A, k, exp in tests:
        b = brute_force(A, k)
        o = optimal(A, k)
        be = best(A, k)
        status = "OK" if b == o == be == exp else f"FAIL(b={b},o={o},be={be})"
        print(f"A={A}, k={k}: {b} (exp={exp}) [{status}]")
